src/Microsoft.AspNetCore.Routing/Matchers/DefaultEndpointSelector.cs

```diff
@@ -11,17 +11,34 @@ namespace Microsoft.AspNetCore.Routing.Matchers
 {
     internal class DefaultEndpointSelector : EndpointSelector
     {
+        private readonly IEndpointSelectorPolicy[] _policies;
+        
+        public DefaultEndpointSelector(IEnumerable<MatcherPolicy> policies)
+        {
+            if (policies == null)
+            {
+                throw new ArgumentNullException(nameof(policies));
+            }
+
+            _policies = policies.OrderBy(p => p.Order).OfType<IEndpointSelectorPolicy>().ToArray();
+        }
+
         public override Task SelectAsync(
             HttpContext httpContext,
             IEndpointFeature feature,
-            CandidateSet candidates)
+            CandidateSet candidateSet)
         {
+            for (var i = 0; i < _policies.Length; i++)
+            {
+                _policies[i].Apply(httpContext, candidateSet);
+            }
```

Comment:
@JamesNK  - I think you're going to be happy about this.

This is a simple way to plug in a DI service that does a pass over the candidates and gets to remove some from consideration.

So, I spent a few hours this morning trying to wire it up so that MVC would replace the `EndpointSelector` wholesale when it detects some `IActionConstraint` implementations, and it was kinda a mess.

The main issues were:
1. Code duplication - I needed to write another thing does what `EndpointSelector` already does
2. Our DI doesn't support decorators

So I think it's probably warranted to step back from that cliff. 

This is going to be very reasonable perf-wise as well since it's using data structures we already have.  We still want to do the graph node thing for perf-sensitive features.

I still want to make MVC able to run `IActionConstraints` for compat. Reaction PR to follow (obviously) since this breaks lots of things. I expect I won't be able to merge this until your CORS work is done.

Reply:
If someone wants per request DI do they access `HttpContext.RequestServices`?

Reply:
> (maybe all together in one back-compat policy that then executes the current action constraint selector)?

Yes, exactly.

> If someone wants per request DI do they access HttpContext.RequestServices?

Yes.

Reply:
> If someone wants per request DI do they access HttpContext.RequestServices?

BTW

The thing about this (pre-rant in case it comes up)... If we did add support for handling per-request instances of things, we could do it by having a factory that calls into the service provider to register it per-request. For all intents and purposes it's equivalent to the user doing it themselves. 

The difference is twofold:
- The people that know they want per-request feel emotionally validated because we thought of them. 
- Another set of people will be very confused because they don't know what our guidance is. They see multiple ways of doing things, and they are unsure of what they are actually supposed to do. 

An example of the latter is filters. We have many ways of plugging in filters. Some of the most relevant ones to this discussion are `[TypeFilter]` and `[ServiceFilter]` both of these allow you to plug in a filter per request, but they do it in different ways.

As a result we get lots of questions about which one to use in which case. It's clear that what we build provides flexibility that folk don't need. You might go so far as to call it *vexibility*, because novices are vexxed.